### DISCUSSION TITLE
Azure key vault integration problem. Invalid tenant id provided.

### DISCUSSION THREAD

**Discussion by KIRY4** [apache/airflow#16219] [Q&A]
Hello everyone,

I have a problem with integration of Airflow and Azure Key Vault. First of all from my point of view documentation regarding this integration bit unclear. I think it will be very useful to add a piece of code as example which will read one secret for testing purposes (I'm not developer so for guys like me it will be really usefull).

Here is documentation: https://airflow.apache.org/docs/apache-airflow-providers-microsoft-azure/stable/secrets-backends/azure-key-vault.html

My environment: Azure AKS 1.18 + Airflow 2.0.2 deployed via helm chart.

I found following link and tried to implement something like that: https://forum.astronomer.io/t/how-to-set-up-azure-keyvault-as-an-external-secrets-backend-on-astronomer/1065

Also I discussed with my colleague code from link above he told me that it is bit outdated and helped me to adopt it. In the end we got following: 
```
from airflow import DAG
from airflow.operators.python import PythonOperator
from datetime import datetime
from airflow.providers.microsoft.azure.secrets.azure_key_vault import AzureKeyVaultBackend


def get_secrets(**kwargs):
    kv_paylife = AzureKeyVaultBackend()
    variable = kv_paylife.get_variable(key=kwargs["var_name"])
    print(variable)


with DAG('test_keyvault_dag', start_date=datetime(2021, 3, 3), schedule_interval=None) as dag:

    test_task = PythonOperator(
        task_id='test-task',
        python_callable=get_secrets,
        op_kwargs={'var_name': 'TEST'}
    )
```

Than I did following:

1) Created azure service principal with get, list permissions on specific key-vault;
2) Added a secret called AIRFLOW-VARIABLES-TEST;
3) Added following env variables:
AZURE_TENANT_ID=something
AZURE_VAULT_URL=https://my-kv.vault.azure.net/ - BTW with slash in the end or not?
AZURE_CLIENT_SECRET=something
AZURE_CLIENT_ID=something
AIRFLOW__SECRETS__BACKEND=airflow.providers.microsoft.azure.secrets.azure_key_vault.AzureKeyVaultBackend
AIRFLOW__SECRETS__BACKEND_KWARGS={"connections_prefix": "AIRFLOW-CONNECTIONS", "variables_prefix": "AIRFLOW-VARIABLES", "vault_url": $AZURE_KEYVAULT_URL}
4) Run DAG.

And here is my problem/question. In DAG logs I'm getting following error:
```
  File "/home/airflow/.local/lib/python3.8/site-packages/azure/identity/_internal/__init__.py", line 39, in validate_tenant_id
    raise ValueError(
ValueError: Invalid tenant id provided. You can locate your tenant id by following the instructions here: https://docs.microsoft.com/partner-center/find-ids-and-domain-names
```
But I'm 99.99% sure that TENANT_ID is correct, checked it several times. Moreover I have only one TENANT_ID in my Azure account.

P.S: And one more question. Is there a setting / parameter which force to use default connections if specific connection was not found in Azure Key Vault. Because when you enable Azure Key Vault backend, default connections stop working.

Any ideas what I did wrong?

**Comment by KIRY4** [comment]
Any ideas?

**Reply by potiuk** [reply]
I believe how it **should** work is that you should not explicitly create and manage the secret backend as an object. If you properly set the variables, the backend object will be created automatically by Airflow and configured using the variables. You do not need to do anything else. If the configuration is correct, you can access your Azure Vault variable automatically via usual Variable.get(VAR_NAME):  https://airflow.apache.org/docs/apache-airflow/stable/concepts/variables.html 

In this case your Python callable should simply look like this:


```
def get_secrets(**kwargs):
    variable = Variable.get_variable(key=kwargs["var_name"])
    print(variable)
```

The way how Airflow instantiates the backend is different that what you do - instead of using empty constructor it will fill the right values from the airflow configuration variables and call the constructor with the parameters expected by the object https://airflow.apache.org/docs/apache-airflow-providers-microsoft-azure/stable/_api/airflow/providers/microsoft/azure/secrets/azure_key_vault/index.html 

In your case you created AzureVaultBackend() with default parameters rather than those configured in Airflow configuration.
